buildroot/share/PlatformIO/variants/MARLIN_ARTILLERY_RUBY/variant.cpp

```diff
@@ -0,0 +1,172 @@
+  PB_0,  //D58/A8
+  PB_1,  //D59/A9
+  PC_0,  //D60/A10
+  PC_1,  //D61/A11
+  PC_2,  //D62/A12
+  PC_3,  //D63/A13
+  PC_4,  //D64/A14
+  PC_5   //D65/A15
+};
+
+#ifdef __cplusplus
+}
+#endif
+
+// ----------------------------------------------------------------------------
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+/**
+  * @brief  System Clock Configuration
+  *         The system Clock is configured as follow :
+  *            System Clock source            = PLL (HSE)
+  *            SYSCLK(Hz)                     = 84000000
+  *            HCLK(Hz)                       = 84000000
+  *            AHB Prescaler                  = 1
+  *            APB1 Prescaler                 = 2
+  *            APB2 Prescaler                 = 1
+  *            HSE Frequency(Hz)              = 8000000
+  *            PLL_M                          = 8
+  *            PLL_N                          = 336
+  *            PLL_P                          = 4
+  *            PLL_Q                          = 7
+  *            VDD(V)                         = 3.3
+  *            Main regulator output voltage  = Scale2 mode
+  *            Flash Latency(WS)              = 2
+  * @param  None
+  * @retval None
+  */
+WEAK void SystemClock_Config(void)
+{
+  RCC_ClkInitTypeDef RCC_ClkInitStruct;
+  RCC_OscInitTypeDef RCC_OscInitStruct;
+
+  /* Enable Power Control clock */
+  __HAL_RCC_PWR_CLK_ENABLE();
+
+  /* The voltage scaling allows optimizing the power consumption when the device is
+     clocked below the maximum system frequency, to update the voltage scaling value
+   regarding system frequency refer to product datasheet.  */
+  __HAL_PWR_VOLTAGESCALING_CONFIG(PWR_REGULATOR_VOLTAGE_SCALE2);
+
+  /* Enable HSI Oscillator and activate PLL with HSI as source */
+  RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_HSE;
+  RCC_OscInitStruct.HSEState = RCC_HSE_ON;
+  RCC_OscInitStruct.PLL.PLLState = RCC_PLL_ON;
+  RCC_OscInitStruct.PLL.PLLSource = RCC_PLLSOURCE_HSE;
+  RCC_OscInitStruct.PLL.PLLM = 8;
+  RCC_OscInitStruct.PLL.PLLN = 336;
+  RCC_OscInitStruct.PLL.PLLP = RCC_PLLP_DIV4;
+  RCC_OscInitStruct.PLL.PLLQ = 7;
+  if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK) {
+    /* Initialization Error */
+    while (1);
+  }
+  /* Select PLL as system clock source and configure the HCLK, PCLK1 and PCLK2
+     clocks dividers */
+  RCC_ClkInitStruct.ClockType = (RCC_CLOCKTYPE_SYSCLK | RCC_CLOCKTYPE_HCLK | RCC_CLOCKTYPE_PCLK1 | RCC_CLOCKTYPE_PCLK2);
+  RCC_ClkInitStruct.SYSCLKSource = RCC_SYSCLKSOURCE_PLLCLK;
+  RCC_ClkInitStruct.AHBCLKDivider = RCC_SYSCLK_DIV1;
+  RCC_ClkInitStruct.APB1CLKDivider = RCC_HCLK_DIV2;
+  RCC_ClkInitStruct.APB2CLKDivider = RCC_HCLK_DIV1;
+  if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_2) != HAL_OK) {
+    /* Initialization Error */
+    while (1);
+  }
+}
+
+#ifdef __cplusplus
+}
+#endif
+
+void flashFirmware(const int16_t) {
+  *((unsigned long *)0x2000FFF0) = 0xDEADBEEF; // End of RAM
+  NVIC_SystemReset();
+}
```

Comment:
This is what is causing the error @gzhuli since it's already defined / defined differently in the HAL:

https://github.com/MarlinFirmware/Marlin/blob/2c8dce69e16daa08de4a62f115b06fe68c0db17d/Marlin/src/HAL/STM32/HAL.cpp#L156-L157

Reply:
Need to add WEAK prefix to flashFirmware() function in src/HAL/STM32/HAL.cpp. The motherboard redefined the flashFirmware() function to issue a special reboot sequence to enter DFU mode when M997 is issued.

https://github.com/artillery3d/Marlin/blob/6eff9aec1a2d375ccb4010896ce02c3969d642c2/Marlin/src/HAL/STM32/HAL.cpp#L156-L157